### PR TITLE
k8s Monitor: Add SandboxID fix.

### PR DESCRIPTION
**Pull request by abhijitherekar**
#### Description
Adds a SandboxID extractor for the puID.

#### Test plan
*Outline the test plan used to test this change before merging it.*

> Fixes #.

### CODE REVIEW

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "fa2f726462fde9b3801b3a8501d0ae61a2f0f32b"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "63d1bec95152059508d5190ad55c8de785a0cbff"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "fa2f726462fde9b3801b3a8501d0ae61a2f0f32b"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "09cda457c154c14a4aaba11f351252735af1448f"
  }
]
```

File: monitor/internal/pod/controller.go
Comment (mheese) [inline monitor/internal/pod/controller.go]: this is problematic, the pod phases never match 100% what the status of containers is. In fact, when init containers come up, the pod phase will still be pending, but we can already extract the sandbox ID. That's actually the reason why I'm using that dance of testing if a container is started below. There might also be a condition in the pod status which actually does exactly what we want. That is worth investigating.

Also, let's not forget, that in 3.14 we also want to have an alternate implementation using the CRI, which will know the sandbox ID already before any init container has been started. We also need to try to keep this friendly enough that we can keep this "user-friendly" in the monitor here. The CRI stuff will most likely remain in trireme, while we pass this extractor here from the enforcer.

To solve your issue here, I would do the following: try to get the sandbox ID without checking for any phase (basically just remove 148), and if it fails, reconcile with an error - not the end of the world and we should cover all our bases here.

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: I wonder ... what does this mean in the case that the pod and/or sandbox are already gone?

Comment (mheese) [reply monitor/internal/pod/delete_controller.go]: I think we'll probably have to make that check now a triple: namespacedName && puID && sandboxID 

However, I might be wrong :) ... Just food for thought

Comment (abhijitherekar) [reply monitor/internal/pod/delete_controller.go]: Awesome point, let me think over it and make changes.

File: monitor/internal/pod/controller.go
Comment (abhijitherekar) [reply monitor/internal/pod/controller.go]: hmm, will make the changes accordingly. But that makes sense if not able to find the snadboxID then reconcile again.

File: controller/internal/processmon/processmon.go
Comment (mheese) [inline controller/internal/processmon/processmon.go]: also please remove before we merge :)

File: monitor/extractors/interface.go
Comment (mheese) [inline monitor/extractors/interface.go]: typo: `sanboxID` ... as this is just a type, you can just say `(string, error)`

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: this is a pointer on purpose and must be a pointer: we don't want to pass copies of this map

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: you are already using the nativeID as key ... not really necessary to have it in here as well, no?

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: now we are missing early handling of the delete that you deleted above.

All in all if there is a pod in the API, we need to treat these two cases:
1. if the namespaced names are the same but the pod UID is different, destroy the PU (the case missing)
2. if the namespaced names are the same, and the pod UIDs are the same, but the the sandbox IDs differ, destroy the PU (the one you do below)

and I think what you are doing in 148 is correct ... if we can't extract the sandbox ID, we don't know what to do... however, how does that look like for stopped PUs? I believe it is still correct: we stop PUs that are in a stopped state, so they will be "initialized" PUs for us, we don't want to destroy them... please think this through as well

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: I usually just construct it from a new struct to avoid having to deal with the error that seems pointless :)

File: monitor/internal/pod/delete_controller.go
Comment (abhijitherekar) [reply monitor/internal/pod/delete_controller.go]: I thought in go the maps are passed as a pointer/by reference by default. 
The compiler usually takes care of passing the reference to the next function.

File: monitor/internal/pod/delete_controller.go
Comment (abhijitherekar) [reply monitor/internal/pod/delete_controller.go]: Yup, good point. WIll make changes.

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: you already got the delObj at the beginning of the function, no need for a 2nd retrieval and error check

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: That's at most info level. It is definitely not an error.

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: typo "destoryed"

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go:183]: add to the comment that we really only update the object if we have never extracted a sandbox before... that's key to the whole treatment here

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: can we rename nativeID to podUID?

we've been changing forth and back here so much ... and my gut tells me that we might not even be at the end of it ... so I'd like to be explicit (even though this is kind of nitpicking ;) )

File: monitor/internal/pod/controller.go
Comment (mheese) [inline monitor/internal/pod/controller.go]: Can we also rename here NativeID to PodUID? And Maybe Key to NamespacedName?

Again, just for clarity in the future as we've been going forth and back on the NativeIDs

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: please turn into Debug

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "3d807212f6b83509cdadd8fb34963eb208a80a99"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "0c829082f5841d76ad9704976aa8c6127b9132f7"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "3d807212f6b83509cdadd8fb34963eb208a80a99"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "05fef40ae41f18ef07938c4786096bab9a169b76"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "3d807212f6b83509cdadd8fb34963eb208a80a99"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "bf7f1bf7dc8459bf793590c078ed30ca85ff57b5"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "c7eaec7c8e9d71fa9a6a985ee4da9da21cdb04c4"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "3d807212f6b83509cdadd8fb34963eb208a80a99"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "d6b7bd4266483ae151a2889bd8c08d85d7f26dc9"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "c7eaec7c8e9d71fa9a6a985ee4da9da21cdb04c4"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "6c45fa2ea854e120ab66c8e7a18f29e7a30fb8ce"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "c7eaec7c8e9d71fa9a6a985ee4da9da21cdb04c4"
  }
]
```

File: monitor/internal/pod/controller.go
Comment (mheese) [inline monitor/internal/pod/controller.go]: again, this can happen on a regular basis ... just change to Debug log please

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: just writing this once here: can we prefix all `zap.L().` logs with `deleteController: ` in the `delete_controller.go` please?

some of those messages are similar to messages from the normal controller ... and it makes it easier to pinpoint

**Review by mheese — CHANGES_REQUESTED** [review]
Please change the log levels as mentioned in my last two comments.

As we have passed FT right now in https://ci.aporeto.io/teams/main/pipelines/pr-tests/jobs/functional-tests/builds/191 already, I'm happy to merge this without a rerun if this is all that needs changing.

File: monitor/internal/pod/delete_controller.go
Comment (mheese) [inline monitor/internal/pod/delete_controller.go]: one last log change: again this can happen regularly, so let's make this debug please

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "K8s_SandboxID_fix",
    "component": "enforcerd",
    "pr-id": "1339",
    "commit-sha": "42ffa630e9e31b4a1cea3f9449054886539669e5"
  },
  {
    "project": "K8s_SandboxID_fix",
    "component": "trireme-lib",
    "pr-id": "856",
    "commit-sha": "43cf92aad11a1fd0d7349b3bee48aa3dcda54750"
  }
]
```

**Review by mheese — APPROVED** [review]
lgtm

**Comment by mheese** [comment]
FTs that are relevant have passed (https://ci.aporeto.io/teams/main/pipelines/pr-tests/jobs/functional-tests/builds/210), merging this now

**Pull request merged**
